design/design-doc-template-stacks-experience.md

```diff
@@ -0,0 +1,505 @@
+
+#### Create
+
+Instantiation is also very similar to any stack:
+
+```
+cat > my-instance.yaml <<EOF
+apiVersion: thing.samples.stacks.crossplane.io/v1alpha1
+kind: ThingInstance
+metadata:
+  name: thinginstance-sample
+spec:
+  myfield: myvalue
+EOF
+
+kubectl apply -f my-instance.yaml
+```
+
+The difference is that underneath, the ThingInstance will be used as the
+input for rendering a template.
+
+#### Delete
+
+When the ThingInstance is deleted, the corresponding resources will also
+be deleted. This is the same behavior as in any stack.
+
+```
+cat > my-instance.yaml <<EOF
+apiVersion: thing.samples.stacks.crossplane.io/v1alpha1
+kind: ThingInstance
+metadata:
+  name: thinginstance-sample
+spec:
+  myfield: myvalue
+EOF
+
+kubectl delete -f my-instance.yaml
+```
+
+The instance is also deleted if the stack is uninstalled.
+
+#### Uninstall
+
+Uninstall looks the same as any other stack:
+
+```
+kubectl crossplane stack uninstall mystack
+```
+
+#### Upgrade and update
+
+Upgrading a stack version, and updating an instance of an object managed
+by the stack, are out of scope of this document. However, one could
+imagine that the process of upgrading a stack version would change a
+version number on object instances which go with the stack, and that the
+version number change would underneath cause the template to be rendered
+again and applied to the cluster.
+
+### The stack.yaml
+
+Configuration for the template stack will be in a `stack.yaml` file at
+the build root of the stack in the repository. For most cases, this will
+mean the `stack.yaml` lives at the root of the repository.
+
+#### Specifying how to process render requests
+
+A Stack author must create a `stack.yaml` in order to configure how
+templates are rendered in response to a render request. Here's an
+example, with comments explaining the directives:
+
+```
+cat > stack.yaml <<EOF
+# This field configures which templates are rendered in response
+# to a given object type. An instance of the object type is
+# considered to be a "render request".
+behaviors:
+  crds:
+    # This is a particular CRD which is being configured. When the
+    # controller sees an object of this type, it will do something.
+    wordpressinstance.wordpress.samples.stacks.crossplane.io:
+      # This is a top-level object to namespace hook configurations.
+      # There can be hooks for multiple different types of events.
+      hooks:
+        # Post create is triggered after an instance is created
+        postCreate:
+          # These are the templates which should be rendered when an object
+          # of the type above is seen.
+          #
+          # Defaults for the variables can be defined in the default values
+          # for the CRD fields, using the standard Kubernetes mechanism for
+          # specifying CRD field default values.
+          # Note that this is a list of objects, so multiple can be
+          # specified.
+          - directory: wordpress
```

Comment:
is specifying the "engine" needed, or can that be inferred by the template controller by looking at the given directory and examining it's structure? (i.e. this dir has a kustomization.yaml so I should render with kustomize)

Reply:
I've added some words to clarify this, so I'm going to resolve this thread.